### PR TITLE
Add workflow for internal releases

### PR DESCRIPTION
**Pull request by VersusFacit**
resolves  38 in epic

### Description

The workflow that powers releases internally. Thoroughly tested using Redshift. Annotations in code for your reading pleasure.

Also, the target build scheme was modified ever so slightly. 

`+buildxxx` where `xxx` is a natural number -- simple and compliant with code artifacts very very specific wants for version names.

### Checklist

- [X] I have read [the contributing guide](https://github.com/dbt-labs/dbt-release/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [X] I have run this code in development and it appears to resolve the stated issue

### CODE REVIEW

File: .github/workflows/internal-archive-release.yml [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        run: |
+          echo Warehouse name:                     ${{ inputs.dbms_name }}
+          echo The release version number:         ${{ inputs.version_number }}
+          echo The release sha:                    ${{ inputs.sha }}
+          echo Package test command:               ${{ inputs.package_test_command }}
+
+      - name: "Checkout main branch"
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.sha || 'main' }}
```
Comment (VersusFacit) [inline .github/workflows/internal-archive-release.yml:53]: Works with a branch or a raw sha

File: .github/workflows/internal-archive-release.yml [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        run: |
+          python -m pip install --user --upgrade pip
+          python -m pip install tox
+          python -m pip --version
+          python -m tox --version
+
+      - name: "Run Tox"
+        run: tox
+
+  run-integration-tests:
```
Comment (VersusFacit) [inline .github/workflows/internal-archive-release.yml:99]: May require some additional setup in each repo's secrets.

File: .github/workflows/internal-archive-release.yml [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+          aws configure set region ${{ secrets.AWS_REGION }} --profile ${{ secrets.AWS_PROFILE_NAME }}
+          aws configure set output text --profile ${{ secrets.AWS_PROFILE_NAME }}
+          aws codeartifact login --tool twine --repository ${{ secrets.AWS_REPOSITORY }} --domain ${{ secrets.AWS_DOMAIN }} --domain-owner ${{ secrets.AWS_DOMAIN_OWNER }} --region ${{ secrets.AWS_REGION }} --profile ${{ secrets.AWS_PROFILE_NAME }}
+
+      - name: "Alter version in metadata of python package"
+        run: |
+          version_file="dbt/adapters/${{ inputs.dbms_name }}/__version__.py"
+          version_in_file=$(grep 'version =' "${version_file}" | cut -d '"' -f2)
+
+          versions_on_aws="$(aws codeartifact list-package-versions --repository ${{ secrets.AWS_REPOSITORY }} --domain ${{ secrets.AWS_DOMAIN }} --domain-owner ${{ secrets.AWS_DOMAIN_OWNER }} --region ${{ secrets.AWS_REGION }} --profile ${{ secrets.AWS_PROFILE_NAME }} --format pypi --package dbt-${{ inputs.dbms_name }} --output json  --query 'versions[*].version' | jq -r '.[]' | grep "^${{ inputs.version_number }}" || true )"  # suppress pipefail only here
```
Comment (VersusFacit) [inline .github/workflows/internal-archive-release.yml:201]: Yes grep does this. No I don't like doing this.

File: .github/workflows/internal-archive-release.yml [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+          current_latest_version="$(echo "${versions_on_aws}" | sort -V | tail -n 1 )"
+
+          echo "[Debug] version_in_file: ${version_in_file}"
+          echo "[Debug] current_latest_version: ${current_latest_version}"
+
+          echo ">>> Altering ${version_file}"
+          if [[ -z "${current_latest_version}" ]]; then
```
Comment (VersusFacit) [inline .github/workflows/internal-archive-release.yml:209]: Smart enough to handle cases where no version yet exists so it won't just jump to special build numbers.

File: .github/workflows/internal-archive-release.yml [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+          sudo apt-get update
+          sudo apt-get install libsasl2-dev
+
+      - name: "Set up Python - ${{ env.PYTHON_TARGET_VERSION }}"
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ env.PYTHON_TARGET_VERSION }}
+
+      - name: "Set Up Postgres (Linux)"
+        if: ${{ contains(github.repository, 'dbt-labs/dbt-core') }}
```
Comment (emmyoop) [inline .github/workflows/internal-archive-release.yml:153]: Should this point to dbt-postgres now?  Or is it just not needed?

File: .github/workflows/internal-archive-release.yml [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        uses: ./.github/actions/setup-postgres-linux
+
+      - name: "Install python tools"
+        run: |
+          python -m pip install --user --upgrade pip
+          python -m pip --version
+          python -m pip install tox
+          tox --version
+
+      - name: Run tests
```
Comment (emmyoop) [inline .github/workflows/internal-archive-release.yml:163]:
```suggestion
      - name: "Run tests"
```

File: .github/workflows/internal-archive-release.yml [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+        run: tox
+
+  create-internal-release:
+    name: Create release for internal archive
+    runs-on: ubuntu-latest
+    needs: [run-unit-tests, run-integration-tests]
+
+    steps:
+      - name: "Checkout main branch"
```
Comment (emmyoop) [inline .github/workflows/internal-archive-release.yml:172]: Same as above

File: .github/workflows/internal-archive-release.yml [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+          echo ">>> Altering ${version_file}"
+          if [[ -z "${current_latest_version}" ]]; then
+              echo "Skipping as no version on aws yet!"
+          elif [[ ${current_latest_version} =~ (.*build)([0-9]+)$ ]]; then
+              base="${BASH_REMATCH[1]}"
+              number="${BASH_REMATCH[2]}"
+              new_number=$((number + 1))
+              tee <<< "version = \"${base}${new_number}\"" "${version_file}"
+          else
```
Comment (emmyoop) [inline .github/workflows/internal-archive-release.yml:216]: Is the `else` just an extra safety step?  I'm not following how we would expect to have versions in aws but they're in the wrong format.

File: .github/workflows/internal-archive-release.yml [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        run: |
+          check-wheel-contents dist/*.whl --ignore W007,W008
+
+      - name: "Upload Build Artifact - ${{ inputs.version_number }}"
+        run: |
+          twine upload --repository codeartifact dist/*
+
+          version_file="$(echo "dbt/adapters/${{ inputs.dbms_name }}/__version__.py")"
+          version="$(grep 'version =' "${version_file}" | cut -d '"' -f2)"
+          echo "-- Success -- released ${version}"
```
Comment (emmyoop) [inline .github/workflows/internal-archive-release.yml:240]: Similar to above, I would make this a notification so it's easy to see on the workflow run Summary page.  [notice example](https://github.com/dbt-labs/actions/blob/92e1b47e7ae44730bd2dd45665219d00b61c2d87/.github/workflows/release-branch-tests.yml#L75-L79)

**Review by emmyoop — CHANGES_REQUESTED** [review]
A few questions but mostly just some formatting changes/clarifications.

File: .github/workflows/internal-archive-release.yml (hunk 8 above)
Comment (VersusFacit) [reply .github/workflows/internal-archive-release.yml:216]: Without this, you skip the bare package and so technically it's a matter of propriety. "Send up the bare patch, then start sending our patches."

File: .github/workflows/internal-archive-release.yml [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+          python -m pip install --user --upgrade pip
+          python -m pip install tox
+          python -m pip --version
+          python -m tox --version
+
+      - name: "Run Tox"
+        run: tox
+
+  run-integration-tests:
+    runs-on: ubuntu-20.04
```
[the remark this answers is not in the document]
Comment (VersusFacit) [reply .github/workflows/internal-archive-release.yml:100]: I think because of convention. I'll change back if it's a problem.

**Review by emmyoop — APPROVED** [review]
Approving as is for now with the understanding that more PRs will follow on after testing since none of this can be tested before the workflow gets merged.  Since this is a workflow that will be manually submitted there is little risk in this approach.
